server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -179,7 +179,20 @@ public void onRejection(Exception e) {
                             e, primary, docWriteRequest.opType() == DocWriteRequest.OpType.DELETE, docWriteRequest.version()),
                         context, null);
                 }
-                finishRequest();
+
+                // Force the execution to finish the request
+                executor.execute(new ActionRunnable<>(listener) {
+
+                    @Override
+                    protected void doRun() {
+                        finishRequest();
```

Comment:
I was thinking that rather than scheduling only `finishRequest` on the write thread pool, we should execute the entirety of `onRejection` on the write thread pool (and still forcing execution, since rejecting execution here would be very bad): https://github.com/elastic/elasticsearch/blob/3ad8aa6d46580229823d298e9ce68ba3aaadc3d6/server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java#L160

The reason I think this is because we're still doing work in `onRejection` that is linear in the number of documents in the bulk request. It seems we'd want to get that off of the networking/cluster state applier thread too given that we're going to fork anyway.

Reply:
I wonder if we should instead let `onRejection` force-push the `onRejection` handling onto the queue of the requested executor. The current "direct" handling kind of prioritizes the `onRejection` handling over everything else in the queue, which I think there is really no good reason for.

Looking at various `onRejection` handlers, some call `listener.onFailure` and verifying that none of those do bad things is tricky.

Also notice that `onAfter` is also called in the caller thread when requests are rejected, this poses similar issues (not that I found a smoking gun there).

Finally, notice that `AbstractRunnable.onRejection` by default calls `onFailure`.

I doubt that we careful consider that `onAfter` and `onFailure` might run in the current thread when using `AbstractRunnable` and executing the `onRejection` handling on the target thread-pool would fix this, making it easier to reason about.

That said, I think this PR is good and I am not objecting to it going in. Following my suggestion is likely to surface a few additional things to resolve.

Reply:
I made @jasontedor change.

I think maybe Henning's comment about maybe `onRejection` should be executed on the thread pool anyway is beyond the scope of this PR? Or at least a larger discussion.